File: cpp/61. Rotate List.cpp
/*
2016-10-2 00:43:05

https://leetcode.com/problems/rotate-list/

Given a list, rotate the list to the right by k places, where k is non-negative.

For example:
Given 1->2->3->4->5->NULL and k = 2,
return 4->5->1->2->3->NULL.

*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        
		ListNode* tmp = head;
		
		int length = 0;
		
		while (tmp != nullptr)
		{
			++ length;
			tmp = tmp->next;
		}
		
		if (length == 0 || k % length == 0)
		{
			return head;
		}
		
		k %= length;
		
		k = length - k;
		
		tmp = head;
		
		for (int i = 1; i < k; ++ i)
		{
			tmp = tmp->next;
		}
				
		ListNode* res = tmp->next;
		tmp->next = nullptr;
		
		tmp = res;
		
			
		while (tmp->next != nullptr)
		{
			tmp = tmp->next;
		}
		
		tmp->next = head;
		
		return res;
		
    }
};

File: java/src/main/java/club/cser/leetcode/61. Rotate List.java
package club.cser.leetcode;

class RotateList {
    public ListNode rotateRight(ListNode head, int k) {

        if (head == null || head.next == null || k == 0)
            return head;

        ListNode cur = head;

        // cur走k步
        while (k > 0) {
            -- k;
            if (cur.next != null) {
                cur = cur.next;
            } else {
                cur = head;
            }
        }

        if (cur == head)
            return head;

        //找到距离尾部k个节点的位置
        ListNode res = head;
        while (cur.next != null) {
            res = res.next;
            cur = cur.next;
        }

        // 尾部接到头部
        cur.next = head;
        // 断开
        cur = res.next;
        res.next = null;

        return cur;

    }
}